### ISSUE TITLE
Improving requirements and dependency check for deploying

### ISSUE THREAD

**Issue by vpenades** [microsoft/onnxruntime#14064]
### Describe the issue

We're working on a product using ORT that is to be deployed to clients in the wild, as discussed in #13276, when deploying in the wild it's not possible to know in advance what's installed or not in the machine before running.

I am aware that DirectML and CPU providers require a certain version of VCRedist to be installed in the machine in order to run.

I need to add the VCRedit version check to my **installer**, so if the VC redist is missing, prompt the user to install the required runtimes... _So I need to know which is the exact minimum version required by the providers to run_.

I looked around other issues, pointing to [https://github.com/Microsoft/onnxruntime#system-requirements](https://github.com/Microsoft/onnxruntime#system-requirements) but that link is broken.

### To reproduce

Just try to use OnnxRuntime in a brand new machine, 5 minutes after installing Windows OS, and you'll see how many things fail...

As it happens, most of our clients are buying new machines just to use our software, and most of them are not tech savvy, So we end up having to guide them manually, by phone, to check the errors and install the proper runtimes (Window updates up to the right version, VC Redist, Graphics drivers, CUDA, etc). You don't realize how painful it is until you suffer it.

### Urgency

I am writing the installer right now, so at the very least, I would need to know the VC Redist minimum version used by ONNX

### Platform

Windows

### OS Version

Win10

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

1.13.1

### ONNX Runtime API

C#

### Architecture

X64

### Execution Provider

DirectML

### Execution Provider Library Version

1.13.1

**Comment by yuslepukhin** [comment]
Installing your dependencies side by side is the recommended method. This also applies to DML.

**Comment by vpenades** [comment]
> Installing your dependencies side by side is the recommended method.

Yes, that's what I plan to do, but I need to know which is the version of VC runtime onnx requires, so I can check of its already installed or not.

The question was, which version requires onnxruntime?

**Comment by yuslepukhin** [comment]
See [here](https://learn.microsoft.com/en-us/cpp/windows/latest-supported-vc-redist?view=msvc-170#visual-studio-2015-2017-2019-and-2022) 

You can just get the latest

> Visual Studio versions since Visual Studio 2015 share the same Redistributable files. For example, any apps built by the Visual Studio 2015, 2017, 2019, or 2022 toolsets can use the latest Microsoft Visual C++ Redistributable. However, the version of the Microsoft Visual C++ Redistributable installed on the machine must be the same or higher than the version of the Visual C++ toolset used to create your application.

**Comment by vpenades** [comment]
Okey, I think there's a misunderstanding here.  I know how to find and install the latest VC Redist runtimes, _but that's not the problem_.

Good installation practices require not blindly installing something that is already installed, so the proper way to install VC Redist is to check first if it's already installed and its version is high enough to run OnnxRuntime.

So I need to know which VCRedist version uses OnnxRuntime, to choose whether to run the installer or not.

**Comment by yuslepukhin** [comment]
We will be working on providing a good clear instruction for the issue you raised.

The article I mentioned earlier specifically states that the redistributables are version independent. The Onnxruntime version you are using is built with VS Studio 2019 so you can always use the latest version.

As for the user, you do not need to ask them to install it. You can ship the redistributable binaries and install them side by side with your aplication.

**Comment by vpenades** [comment]
So, knowing that OnnxRuntime is built against VC 2019, and checking against [this table](https://vcredist.com/versions/) then I assume the minimum version OnnxRuntime requires is **14.29.30139.0** 

Since you say that resdistributables are version independent, does it mean that OnnxRuntime could also run with 14.16.27033.0 (VS2017) and 14.0.24215.1 (VS2015) ??

The distinction is important, because it's the difference between needlessly requesting a user to install the redist runtime.

In other words:

```c#

var vcRedistInstalledVersion = DetectVCRedistInstalledVersion();

if ( vcRedistInstalledVersion < 14.0.24215.1 )  // check if we meet OnnxRuntime minimum requirements
{
   RequestUserToInstallRedistRuntime(); // Ask the user to install VC Redist latest version.
}
else
{
  SafelyRunOnnxRuntime(); // good to go!
}
```

Can you confirm this would do?  or should I set the higher **4.29.30139.0** version in the check?

**Comment by yuslepukhin** [comment]
> Since you say 

VC Redist is a system component, and it is governed by the rules that are described in the documentation. Neither I nor ONNXRuntime `have a say` in how it is delivered and updated. It is updated via Windows Update once centrally installed on a box.

> that resdistributables are version independent, does it mean that OnnxRuntime could also run with 14.16.27033.0 (VS2017) and 14.0.24215.1 (VS2015) ??

For exact versioning, please, refer to https://learn.microsoft.com/en-us/cpp/windows/redistributing-visual-cpp-files?view=msvc-170#install-the-redistributable-packages

And https://learn.microsoft.com/en-us/cpp/windows/deploying-visual-cpp-application-by-using-the-vcpp-redistributable-package?source=recommendations&view=msvc-170

Said that, the Universal Runtime has been available starting with VS 2015 and it is binary compatible onwards. You may want to check what to do on the boxes that have VS Runtime from prior versions.

I do not know much about your customers, but on a new box, such as new VM, you could probably just attempt to install the latest.

**Comment by fdwr** [comment]
> I am aware that DirectML and CPU providers require a certain version of VCRedist to be installed in the machine in order to run.

@vpenades Although the DirectML.dll *OS build* depends on the shared CRT DLL (as appropriate for longer term security versioning), the much more regularly updated *redist* is self-contained and shouldn't (AFAICS) require additional CRT installation steps for the customer or system DLL updates.

**Comment by faxu** [comment]
@vpenades the updated link which notes the requirements is here: https://onnxruntime.ai/docs/install/#requirements
"Windows builds require VC Runtime 2019." According to the [visual C++ runtime documentation](https://learn.microsoft.com/en-us/cpp/windows/redistributing-visual-cpp-files?view=msvc-170#install-the-redistributable-packages),14.0 should be compatible with VS 2015, 2017, 2019, and 2022. 

"The version number is 14.0 for Visual Studio 2015, 2017, 2019, and 2022 because the latest Redistributable is binary compatible with previous versions back to 2015."

**Comment by faxu** [comment]
@vpenades, ORT is built using VS 2019, as noted in requirements [here](https://onnxruntime.ai/docs/install/#requirements)

From reviewing the various Visual C++ Runtime documentation, it seems that it's recommended to always install the latest which includes bug fixes, security fixes, etc. If the redist is included in your application installer, Windows Update can also update these packages. (see: [Determine which DLLs to redistribute](https://learn.microsoft.com/en-US/cpp/windows/determining-which-dlls-to-redistribute?view=msvc-170)) Is there an issue with adding this to your application installation to ensure it's installed for all users rather than doing a version check? The installer has logic built in to avoid redundant installs.

**Comment by faxu** [comment]
@vpenades
If you want to avoid running the installer to get the latest version, then you can do the check for existing version as you proposed above. You can find the supported Visual C++ Redist for VS2019 at https://my.visualstudio.com/, search term "Visual C++ Redistributable"

You can also consider using the [/quiet option](https://learn.microsoft.com/en-us/windows/win32/msi/standard-installer-command-line-options#quiet) for the install to avoid showing messages to the user.

**Comment by tildekarthik** [comment]
I am not an expert in dotnet or CPP but here is what solved the problem - I installed Visual C++ runtime for x64 (Server was Xeon running Windows 2016 server 64 bit)

solved the problem for me

**Comment by vpenades** [comment]
I created [this discussion](https://github.com/microsoft/onnxruntime/discussions/15309) to talk about handling dependencies.

The issue includes installing VC Redist but not limited to it, since each ORT has it's unique set of requirements and dependencies.

Also, this issue is probably more meaningful for use cases where the application is being distributed to be installed in end user's machines where you don't have control over the host environment. That's why some of the solutions provided here don't apply.
